src/credentials.rs

```diff
@@ -63,14 +63,22 @@ impl Credentials {
 
     /// Reads the credentials from the file system
     pub async fn read() -> miette::Result<Self> {
-        let raw: RawCredentialCollection = toml::from_str(
-            &fs::read_to_string(location().into_diagnostic()?)
-                .await
-                .into_diagnostic()
-                .wrap_err(ReadError(CREDENTIALS_FILE))?,
-        )
-        .into_diagnostic()
-        .wrap_err(DeserializationError(ManagedFile::Credentials))?;
+        let location = location().into_diagnostic()?;
+
+        // if the file does not exist, we don't need to treat it as an error.
+        let contents = match fs::read_to_string(&location).await {
+            Ok(contents) => contents,
+            Err(error) if error.kind() == ErrorKind::NotFound => Default::default(),
```

Comment:
I disagree with the location of this. I believe you should not change the semantics of `read`.

You should do: `Credentials::read().await.unwrap_or_default()` in the `buffrs::build` script.

For transparency you could use things like `tap` to log out potential errors as cargo warnings. The "Read or default" semantics are really only ever needed in the build script

Reply:
We cannot do what you suggested, `Credentials::read().await.unwrap_or_default()` is incorrect as it would mask critical errors such as permissions issues. That is dangerous, because it means you could be in a situation where a credentials file clearly exists, and the tool throws no error when you run it, but the credentials are not used. This leads to debugging hell and I would try to avoid it.

At the same time I think I understand where you are coming from. I believe you don't want to implicitly return a default here.

So instead, I will make a compromise: I will have the `read()` function return a `Result<Option<Self>>`. That way, we can clearly disambiguate between a critical error as `Err(..)` and the file simply not existing as `Ok(None)`. I believe that this is the cleanest way of doing it anyways because it lets the callee decide how to treat a missing file.

Reply:
Sounds good! I totally agree with this 😊